lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp

```diff
@@ -2659,58 +2670,94 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
-      new SwiftASTContextForExpressions(m_description, typeref_typesystem));
+  std::shared_ptr<SwiftASTContext> swift_ast_sp;
+  if (for_expressions) {
+    // Expression context.
+    if (!target_sp && for_expressions) {
+      LOG_PRINTF(GetLog(LLDBLog::Types), "No target for expression typesystem");
+      return {};
+    }
+    swift_ast_sp.reset(new SwiftASTContextForExpressions(
+        m_description,
+        *llvm::cast<TypeSystemSwiftTypeRef>(&typeref_typesystem)));
+    // This is a scratch AST context, mark it as such.
+    swift_ast_sp->m_is_scratch_context = true;
+
+    swift_ast_sp->GetLanguageOptions().EnableCXXInterop =
+        ShouldEnableCXXInterop(cu);
+
+    if (target_sp->IsEmbeddedSwift())
+      swift_ast_sp->GetLanguageOptions().enableFeature(
+          swift::Feature::Embedded);
+  } else {
+    // Typesystem fallback context.
+    if (!module_sp) {
+      LOG_PRINTF(GetLog(LLDBLog::Types), "No module for fallback typesystem");
+      return {};
+    }
+    swift_ast_sp.reset(static_cast<SwiftASTContext *>(
+        new SwiftASTContextForModule(m_description, typeref_typesystem)));
+    // This is a module AST context, mark it as such.
+    swift_ast_sp->m_is_scratch_context = false;
+    swift_ast_sp->m_module = module_sp.get();
+    swift_ast_sp->GetLanguageOptions().EnableAccessControl = false;
+    swift_ast_sp->GetLanguageOptions().EnableCXXInterop =
+        module_sp->IsSwiftCxxInteropEnabled();
+
+    if (module_sp->IsEmbeddedSwift())
+      swift_ast_sp->GetLanguageOptions().enableFeature(
+          swift::Feature::Embedded);
+  }
+
   auto defer_log = llvm::make_scope_exit(
       [swift_ast_sp] { swift_ast_sp->LogConfiguration(); });
 
   LOG_PRINTF(GetLog(LLDBLog::Types), "(Target)");
   auto logError = [&](const char *message) {
     LOG_PRINTF(GetLog(LLDBLog::Types), "Failed to create scratch context - %s",
                message);
-    if (StreamSP errs_sp = target.GetDebugger().GetAsyncErrorStream())
-      errs_sp->Printf("Cannot create Swift scratch context (%s)", message);
+    if (target_sp)
+      if (StreamSP errs_sp = target_sp->GetDebugger().GetAsyncErrorStream())
+        errs_sp->Printf("Cannot create Swift scratch context (%s)", message);
   };
 
-  ArchSpec arch = target.GetArchitecture();
+  ArchSpec arch;
+  if (target_sp)
+    arch = target_sp->GetArchitecture();
+  else if (module_sp)
+    arch = module_sp->GetArchitecture();
+
   if (!arch.IsValid()) {
     logError("invalid target architecture");
     return {};
   }
 
-  // This is a scratch AST context, mark it as such.
-  swift_ast_sp->m_is_scratch_context = true;
-
-  swift_ast_sp->GetLanguageOptions().EnableCXXInterop =
-      ShouldEnableCXXInterop(cu);
-
-  if (target.IsEmbeddedSwift())
-    swift_ast_sp->GetLanguageOptions().enableFeature(swift::Feature::Embedded);
-
   bool handled_sdk_path = false;
-  const size_t num_images = target.GetImages().GetSize();
+  ModuleList module_module;
+  if (!target_sp)
```

Comment:
We have a lot of code checking if there's no target in this function which didn't exist before. Is this even possible? Could we be instantiating an AST context without a target?

Should we just return early if we don't have a target?

Reply:
This function is called for 
- creating an SwiftASTContextForExpression (with target)
- creating a precise per-module fallback SwiftASTContext for a TypeSystemSwiftTypeRef (no target)